src/main/network/dns.rs

```diff
@@ -0,0 +1,170 @@
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::Write;
+use std::net::Ipv4Addr;
+use std::os::fd::AsRawFd;
+use std::path::PathBuf;
+use std::sync::Arc;
+
+use rustix::fs::MemfdFlags;
+use shadow_shim_helper_rs::HostId;
+
+#[derive(Debug)]
+struct Database {
+    name_index: HashMap<String, Arc<Record>>,
+    addr_index: HashMap<Ipv4Addr, Arc<Record>>,
+}
+
+#[derive(Debug)]
+struct Record {
+    id: HostId,
+    addr: Ipv4Addr,
+    name: String,
+}
+
+#[derive(Debug)]
+pub struct DnsBuilder {
+    db: Database,
+}
+
+impl DnsBuilder {
+    pub fn new() -> Self {
+        Self {
+            db: Database {
+                name_index: HashMap::new(),
+                addr_index: HashMap::new(),
+            },
+        }
+    }
+
+    pub fn register(&mut self, id: HostId, addr: Ipv4Addr, name: String) {
+        if !addr.is_loopback() && !addr.is_unspecified() {
+            let record = Arc::new(Record {
+                id,
+                addr,
+                name: name.clone(),
+            });
+            self.db.name_index.insert(name, record.clone());
+            self.db.addr_index.insert(addr, record);
+        }
+    }
+
+    pub fn into_dns(self) -> anyhow::Result<Dns> {
+        let pid = std::process::id();
+        let name = format!("shadow_dns_hosts_file_{pid}");
+
+        let memfd = rustix::fs::memfd_create(name, MemfdFlags::CLOEXEC)?;
+        let mut file = File::from(memfd);
+
+        // Sort the records to produce deterministic ordering in the hosts file.
+        let mut records: Vec<&Arc<Record>> = self.db.addr_index.values().collect();
+        records.sort_by(|a, b| a.addr.cmp(&b.addr));
+
+        writeln!(file, "127.0.0.1 localhost")?;
+        for record in records.iter() {
+            writeln!(file, "{} {}", record.addr, record.name)?;
+        }
+
+        let path = PathBuf::from(format!("/proc/{pid}/fd/{}", file.as_raw_fd()));
+
+        Ok(Dns {
+            db: self.db,
+            _hosts_file: file,
+            hosts_path: path,
+        })
+    }
+}
+
+impl Default for DnsBuilder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[derive(Debug)]
+pub struct Dns {
+    db: Database,
+    // Keep this handle while Dns is valid to prevent closing the memfd
+    // containing the hosts database in /etc/hosts format.
+    _hosts_file: File,
+    hosts_path: PathBuf,
+}
+
+impl Dns {
+    pub fn addr_to_host_id(&self, addr: Ipv4Addr) -> Option<HostId> {
+        self.db.addr_index.get(&addr).map(|record| record.id)
+    }
+
+    pub fn addr_to_name(&self, addr: Ipv4Addr) -> Option<&String> {
+        self.db.addr_index.get(&addr).map(|record| &record.name)
+    }
+
+    pub fn name_to_addr(&self, name: &str) -> Option<Ipv4Addr> {
+        self.db.name_index.get(name).map(|record| record.addr)
+    }
+
+    pub fn hosts_path(&self) -> &PathBuf {
```

Comment:
nit: maybe return `&Path` instead?

Reply:
I made a change so we're no longer storing the `PathBuf` in the DNS, and so instead we build one on demand and return the owned `PathBuf` to the caller to do whatever they want with.